hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/ProxyRegistry.java

```diff
@@ -50,7 +52,29 @@
     ProxyRegistry(ProxyServiceImpl proxyService, String serviceName) {
         this.proxyService = proxyService;
         this.serviceName = serviceName;
-        this.service = proxyService.nodeEngine.getService(serviceName);
+        this.service = getService(proxyService.nodeEngine, serviceName);
+    }
+
+    /**
+     * Returns the service for the given {@code serviceName} or throws an
+     * exception. This method never returns {@code null}.
+     *
+     * @param nodeEngine  the node engine
+     * @param serviceName the remote service name
+     * @return the service instance
+     * @throws HazelcastException                  if there is no service with the given name
+     * @throws HazelcastInstanceNotActiveException if this instance is shutting down
+     */
+    private RemoteService getService(NodeEngineImpl nodeEngine, String serviceName) {
+        try {
+            return nodeEngine.getService(serviceName);
+        } catch (HazelcastException e) {
+            if (!nodeEngine.isRunning()) {
+                throw new HazelcastInstanceNotActiveException(e.getMessage());
```

Comment:
Hmm, would it be beneficial to keep the stacktrace of the original exception here, by adding the constructor for `new HazelcastInstanceNotActiveException(e)`?

Reply:
I may but I don't see much benefit here. Since `NodeEngineImpl#getService` actually constructs the exception, there isn't much information contained in the exception stack trace. What I am proposing now is:

```
com.hazelcast.core.HazelcastInstanceNotActiveException: HazelcastInstance[[127.0.0.1]:5001] is not active!

	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getService(ProxyRegistry.java:73)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.<init>(ProxyRegistry.java:55)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl$1.createNew(ProxyServiceImpl.java:81)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl$1.createNew(ProxyServiceImpl.java:79)
	at com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent(ConcurrencyUtil.java:107)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getOrCreateRegistry(ProxyServiceImpl.java:138)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:146)
```

What we get if we add the original exception:
```
com.hazelcast.core.HazelcastInstanceNotActiveException: com.hazelcast.spi.exception.RetryableHazelcastException: HazelcastInstance[[127.0.0.1]:5001] is not active!

	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getService(ProxyRegistry.java:73)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.<init>(ProxyRegistry.java:55)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl$1.createNew(ProxyServiceImpl.java:81)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl$1.createNew(ProxyServiceImpl.java:79)
	at com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent(ConcurrencyUtil.java:107)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getOrCreateRegistry(ProxyServiceImpl.java:138)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:146)
Caused by: com.hazelcast.spi.exception.RetryableHazelcastException: HazelcastInstance[[127.0.0.1]:5001] is not active!
	at com.hazelcast.spi.impl.NodeEngineImpl.getService(NodeEngineImpl.java:369)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getService(ProxyRegistry.java:70)
	... 19 more
```

From the first stack trace it is visible that `ProxyRegistry.getService` failed with `HazelcastInstanceNotActiveException`. With the second stack trace, it's the same information but with more stack. WDYT?